sourced/ml/cmd_entries/repos2bow.py

```diff
@@ -0,0 +1,47 @@
+import logging
+from uuid import uuid4
+
+from scipy import sparse
+
+from sourced.ml.extractors import __extractors__
+from sourced.ml.models import BOW, OrderedDocumentFrequencies, DocumentFrequencies
+from sourced.ml.transformers import Engine, UastExtractor, UastDeserializer, Uast2DocFreq, \
+    Uast2TermFreq, HeadFiles, TFIDF
+from sourced.ml.utils import create_engine
+
+
+def repos2bow_entry(args):
+    log = logging.getLogger("repos2bow")
+    engine = create_engine("repos2bow-%s" % uuid4(), **args.__dict__)
+    document_column_name = "repository_id"
+    extractors = [__extractors__[s](
+        args.min_docfreq, **__extractors__[s].get_kwargs_fromcmdline(args))
+        for s in args.feature]
+
+    uast_pipeline = Engine(engine, explain=args.explain) \
+        .link(HeadFiles()) \
+        .link(UastExtractor(languages=args.languages)) \
+        .link(UastDeserializer())
+    df_pipeline = uast_pipeline.link(Uast2DocFreq(extractors, document_column_name))
+    df = df_pipeline.execute()
+    tf_pipeline = uast_pipeline.link(Uast2TermFreq(extractors, document_column_name))
+    tf = tf_pipeline.execute()
+    tfidf = TFIDF(tf=tf, df=df).execute()
+
+    tokens, documents, values = zip(*tfidf.take(tfidf.count()))
+    log.info("Creating index mappings")
+    doc2id = {doc: doc_id for doc_id, doc in enumerate(sorted(list(set(documents))))}
```

Comment:
`sorted` does not need `list` inside.

Besides: why `set`? Is there a way they can contain duplicates?

Reply:
yes, it contain duplicates, because each (tokens[i], documents[i], values[i]) is one matrix element.

Reply:
Is there a risk of explosion here given the size of the matrix? Then better to calc unique elements on Spark

Reply:
I think that not really, because anyway, we need all these values to save it to the disk, so anyway they should fit memory size.

Reply:
We need **indices** of tokens and documents, not their strings. And we need the lists of unique tokens nd documents. This will fit into memory. Individual, unindexed strings will explode and that is expected.